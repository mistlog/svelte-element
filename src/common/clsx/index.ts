/**
 * from https://github.com/lukeed/clsx
 */
function toVal(mix) {
	var k, y, str = '';

	if (typeof mix === 'string' || typeof mix === 'number') {
		str += mix;
	} else if (typeof mix === 'object') {
		if (Array.isArray(mix)) {
			for (k = 0; k < mix.length; k++) {
				if (mix[k]) {
					if (y = toVal(mix[k])) {
						str && (str += ' ');
						str += y;
					}
				}
			}
		} else {
			for (k in mix) {
				if (mix[k]) {
					str && (str += ' ');
					str += k;
				}
			}
		}
	}

	return str;
}

export type ClassValue = ClassArray | ClassDictionary | string | number | null | boolean | undefined;

export interface ClassDictionary {
	[id: string]: any;
}

export interface ClassArray extends Array<ClassValue> { }

export function Î£(...args: ClassArray) {
	var i = 0, tmp, x, str = '';
	while (i < arguments.length) {
		if (tmp = arguments[i++]) {
			if (x = toVal(tmp)) {
				str && (str += ' ');
				str += x
			}
		}
	}
	return str;
}
