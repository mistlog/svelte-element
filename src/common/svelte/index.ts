import { setContext as rawSetContext, getContext as rawGetContext } from "svelte";

export function setContext<T>(key: keyof T, value: T[keyof T]) {
    return rawSetContext(key, value);
}

export function getContext<T>(key: keyof T): T[keyof T] {
    return rawGetContext(key);
}

export type SvelteInternalProps = { $$slots?: any };
