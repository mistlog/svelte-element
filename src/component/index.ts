export { default as Button } from "./button/Button";
export * from "./button/Button";

export { default as ButtonGroup } from "./button/ButtonGroup";

export { default as Row } from "./layout/Row";
export { default as Col } from "./layout/Col";

export { default as Radio } from "./radio/Radio";
export * from "./radio/Radio";
export { default as RadioGroup } from "./radio/RadioGroup";
export * from "./radio/RadioGroup";

export { default as Checkbox } from "./checkbox/Checkbox";
export * from "./checkbox/Checkbox";
export { default as CheckboxGroup } from "./checkbox/CheckboxGroup";
export * from "./checkbox/CheckboxGroup";

export { default as Menu } from "./menu/Menu";
export { default as MenuItem } from "./menu/MenuItem";
export { default as SubMenu } from "./menu/SubMenu";
