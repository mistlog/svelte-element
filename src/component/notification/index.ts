import Notification, { INotificationWillUnmountEvent, NotificationType } from "./Notification";

export interface INotificationConfig {
    title: string;
    message: string;
    duration?: number;
    type?: NotificationType;
}

export function notification(config: INotificationConfig) {
    const { title = "", message = "", duration = 3000, type } = config || {};

    //
    const div = document.createElement("div");
    document.body.appendChild(div);

    const className = ".el-notification";

    // credit: element react
    const instances = document.querySelectorAll(className);
    const lastInstance = instances[instances.length - 1] as HTMLElement;
    const top = lastInstance ? lastInstance.getBoundingClientRect().bottom + 16 : 16;

    let component = Reflect.construct(Notification, [
        {
            target: div,
            props: { top, title, message, type },
            intro: true
        }
    ]);

    //
    const removeNotification = () => {
        if (component) {
            component.$destroy();
            component = null;
        }
    };

    component.$on("close", () => {
        removeNotification();
    });

    component.$on("willUnmount", (e: INotificationWillUnmountEvent) => {
        // credit: element react
        const { height, top } = e.detail;
        const instances = document.querySelectorAll(className);
        for (let i = 0; i < instances.length; ++i) {
            const element = instances[i] as HTMLElement;
            const rect = element.getBoundingClientRect();
            if (rect.top > top) {
                element.style.setProperty("top", (rect.top - height - 16).toString() + "px");
            }
        }
        document.body.removeChild(div);
    });

    setTimeout(() => {
        removeNotification();
    }, duration);
}
