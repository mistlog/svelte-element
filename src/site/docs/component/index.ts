import h1 from "./h1";
import h2 from "./h2";
import h3 from "./h3";

export { h1, h2, h3 };