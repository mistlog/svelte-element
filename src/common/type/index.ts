import { Readable, Writable } from "svelte/store";

export type InputOnChangeHandler = (event: Svelte.ChangeEvent<HTMLInputElement>) => void;
export type ButtonOnClickHandler = (event: Svelte.MouseEvent<HTMLButtonElement, MouseEvent>) => void;

type TypeWithGeneric<T> = Readable<T> | Writable<T>;
export type AutoSubscribe<Type> = Type extends TypeWithGeneric<infer X> ? X : never;

export type IconName = "edit" | "share" | "delete" | "search" | "arrow-left";