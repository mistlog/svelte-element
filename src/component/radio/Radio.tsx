function UI(state: IRadioState, { value, checked }: IRadioProps) {
    <label class={state.style}>
        <span class={state.inputWrapperStyle}>
            <span class={state.inputSymbolStyle} />
            <input type="radio" checked={checked} class={state.inputStyle} value={value} onChange />
        </span>
        <span class={state.inputLabelStyle}>
            <slot />
        </span>
    </label>
}

export interface IRadioProps {
    /**
     * checked state of radio
     * @default false
     */
    checked?: boolean
    /**
     * whether Radio is disabled
     * @default false
     */
    disabled?: boolean
    /**
     * the value of radio
     */
    value?: string
    onChange?: InputOnChangeHandler
}

export default function _(props: IRadioProps) {
    const { checked = false, disabled = false, value = "" } = props;

    //@ts-ignore
    <State />;

    //@ts-ignore
    <UI />;
}

interface IRadioState {
    inputStyle: string
    inputSymbolStyle: string
    inputWrapperStyle: string
    inputLabelStyle: string
    style: string;
}

function State({ checked, disabled }: IRadioProps) {

    let state = {} as IRadioState;

    state.inputSymbolStyle = "el-radio__inner";
    state.inputStyle = "el-radio__original";
    state.inputLabelStyle = "el-radio__label";
    state.style = "el-radio";

    {
        "use watch";
        state.inputWrapperStyle = Σ(
            "el-radio__input",
            {
                "is-checked": checked,
                "is-disabled": disabled
            }
        );
    }
}

import { Σ } from "../../common/clsx/index";
import { InputOnChangeHandler } from "../../common/type";