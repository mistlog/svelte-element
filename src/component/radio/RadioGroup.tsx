function UI(
    state: IRadioGroupState,
    { radios }: IRadioGroupProps,
    onChangeHandler: InputOnChangeHandler
) {
    <each from={radios}>
        {(radio: IRadio, key = radio.value) => (
            <Radio
                value={radio.value}
                checked={radio.value === state.selected}
                onChange={onChangeHandler}
            >
                {radio.label}
            </Radio>
        )}
    </each>;
}

export interface IRadio {
    value: string;
    label: string;
}

export interface IRadioGroupProps {
    radios: Array<IRadio>;
    defaultSelected?: string;
    /**
     * triggers when the selected radio changes
     */
    onChange?: RadioGroupOnChangeHandler;
}

export default function _(props: IRadioGroupProps) {
    const { radios, defaultSelected } = props;

    //@ts-ignore
    <State />;

    //@ts-ignore
    <Event />;

    //@ts-ignore
    <UI />;
}

//
interface IRadioGroupState {
    selected: string;
}

function State({ defaultSelected }: IRadioGroupProps) {
    const state: IRadioGroupState = { selected: defaultSelected };
}

//
export type RadioGroupEvent = { detail: { selected: string } };
export type RadioGroupOnChangeHandler = (event: RadioGroupEvent) => void;

function Event(state: IRadioGroupState) {
    const dispatch = createEventDispatcher();
    const onChangeHandler: InputOnChangeHandler = event => {
        state.selected = event.target.value;
        dispatch("change", { selected: state.selected });
    };
}

import { default as Radio } from "./Radio";
import { InputOnChangeHandler } from "../../common/type";
import { createEventDispatcher } from "svelte";
