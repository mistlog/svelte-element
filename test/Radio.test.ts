import { render, fireEvent } from "@testing-library/svelte";
import { normalize } from "./common/utility";

import { IRadioProps, IRadioGroupProps } from "../src/component";
import Radio from "../src/component/radio/Radio.svelte";
import RadioGroup from "../src/component/radio/RadioGroup.svelte";

import DefaultRadio from "./assets/DefaultRadio.svelte";
import ClickRadio from "./assets/ClickRadio.svelte";
import TwoRadios from "./assets/TwoRadios.svelte";

describe("Radio", () => {
    test("text", () => {
        const { container } = render(DefaultRadio);
        const radio = container.querySelector(".el-radio");
        expect(normalize(radio.outerHTML)).toMatchSnapshot();
    });

    test("checked", () => {
        const { container } = render(Radio, { checked: true } as IRadioProps);
        const input = container.querySelector(".el-radio__input");
        expect(input.className.includes("is-checked")).toEqual(true);
    });

    test("disabled", () => {
        const { container } = render(Radio, { disabled: true } as IRadioProps);
        const input = container.querySelector(".el-radio__input");
        expect(input.className.includes("is-disabled")).toEqual(true);
    });

    test("click event", async () => {
        const { container } = render(ClickRadio);
        const [a, b] = container.getElementsByTagName("input");
        expect(a.checked).toEqual(true);
        expect(b.checked).toEqual(false);

        await fireEvent(b, new MouseEvent("click", {}));
        expect(a.checked).toEqual(false);
        expect(b.checked).toEqual(true);
    });
});

describe("RadioGroup", () => {
    test("empty radio group", () => {
        const { container } = render(RadioGroup, { radios: [] } as IRadioGroupProps);
        expect(normalize(container.outerHTML)).toMatchSnapshot();
    });

    test("radio group with radios", async () => {
        const { container } = render(TwoRadios);
        expect(normalize(container.outerHTML)).toMatchSnapshot();

        const [a, b] = container.getElementsByTagName("input");
        await fireEvent(b, new MouseEvent("click", {}));
        expect(normalize(container.outerHTML)).toMatchSnapshot();
    });
});
