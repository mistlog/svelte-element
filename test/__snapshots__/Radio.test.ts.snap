// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Radio text 1`] = `
"<label class=\\"el-radio\\">
  <span class=\\"el-radio__input\\">
    <span class=\\"el-radio__inner\\"></span>
    <input type=\\"radio\\" class=\\"el-radio__original\\" value=\\"\\" />
  </span>
  <span class=\\"el-radio__label\\">Option A</span>
</label>
"
`;

exports[`RadioGroup empty radio group 1`] = `
"<body>
  <div></div>
</body>
"
`;

exports[`RadioGroup radio group with radios 1`] = `
"<body>
  <div>
    <label class=\\"el-radio\\">
      <span class=\\"el-radio__input is-checked\\">
        <span class=\\"el-radio__inner\\"></span>
        <input type=\\"radio\\" class=\\"el-radio__original\\" value=\\"a\\" />
      </span>
      <span class=\\"el-radio__label\\">Option A</span>
    </label>
    <label class=\\"el-radio\\">
      <span class=\\"el-radio__input\\">
        <span class=\\"el-radio__inner\\"></span>
        <input type=\\"radio\\" class=\\"el-radio__original\\" value=\\"b\\" />
      </span>
      <span class=\\"el-radio__label\\">Option B</span>
    </label>
  </div>
</body>
"
`;

exports[`RadioGroup radio group with radios 2`] = `
"<body>
  <div>
    <label class=\\"el-radio\\">
      <span class=\\"el-radio__input\\">
        <span class=\\"el-radio__inner\\"></span>
        <input type=\\"radio\\" class=\\"el-radio__original\\" value=\\"a\\" />
      </span>
      <span class=\\"el-radio__label\\">Option A</span>
    </label>
    <label class=\\"el-radio\\">
      <span class=\\"el-radio__input is-checked\\">
        <span class=\\"el-radio__inner\\"></span>
        <input type=\\"radio\\" class=\\"el-radio__original\\" value=\\"b\\" />
      </span>
      <span class=\\"el-radio__label\\">Option B</span>
    </label>
  </div>
</body>
"
`;
